tools/state-viewer/src/epoch_info.rs

```diff
@@ -86,58 +86,56 @@ fn display_block_and_chunk_producers(
-        get_block_height_range(epoch_info, chain_store, epoch_manager)?;
+        get_block_height_range(epoch_id, chain_store, epoch_manager)?;
     let num_shards = epoch_manager.num_shards(epoch_id).unwrap();
     for block_height in block_height_range {
         let bp = epoch_info.sample_block_producer(block_height);
         let bp = epoch_info.get_validator(bp).account_id().clone();
-        let cps: Vec<AccountId> = (0..num_shards)
+        let cps: Vec<String> = (0..num_shards)
             .map(|shard_id| {
                 let cp = epoch_info.sample_chunk_producer(block_height, shard_id);
                 let cp = epoch_info.get_validator(cp).account_id().clone();
-                cp
+                cp.as_str().to_string()
             })
             .collect();
-        println!(
-            "Block height: {}. Block Producer: {}. Chunk Producers: {:?}",
-            block_height, bp, cps
-        );
+        println!("{block_height}: BP=\"{bp}\" CP={cps:?}");
     }
     Ok(())
 }
 
 // Iterate over each epoch starting from the head. Find the requested epoch and its previous epoch
 // and use that to determine the block range corresponding to the epoch.
 fn get_block_height_range(
-    epoch_info: &EpochInfo,
+    epoch_id: &EpochId,
     chain_store: &ChainStore,
     epoch_manager: &EpochManagerHandle,
-) -> Result<Range<BlockHeight>, anyhow::Error> {
+) -> anyhow::Result<Range<BlockHeight>> {
     let head = chain_store.head()?;
     let mut cur_block_info = epoch_manager.get_block_info(&head.last_block_hash)?;
     loop {
-        let cur_epoch_info = epoch_manager.get_epoch_info(cur_block_info.epoch_id())?;
-        let cur_epoch_height = cur_epoch_info.epoch_height();
-        assert!(
-            cur_epoch_height >= epoch_info.epoch_height(),
-            "cur_block_info: {:#?}, epoch_info.epoch_height: {}",
-            cur_block_info,
-            epoch_info.epoch_height()
-        );
-        let epoch_first_block_info =
-            epoch_manager.get_block_info(cur_block_info.epoch_first_block())?;
-        let prev_epoch_last_block_info =
-            epoch_manager.get_block_info(epoch_first_block_info.prev_hash())?;
-        let cur_epoch_start_height = epoch_manager.get_epoch_start_height(cur_block_info.hash())?;
-        let cur_epoch_id = cur_block_info.epoch_id();
-        let next_epoch_start_height =
-            cur_epoch_start_height + epoch_manager.get_epoch_config(cur_epoch_id)?.epoch_length;
-        if cur_epoch_height == epoch_info.epoch_height() {
-            return Ok(cur_epoch_start_height..next_epoch_start_height);
+        if cur_block_info.epoch_id() == epoch_id {
+            let epoch_start_height = epoch_manager.get_epoch_start_height(cur_block_info.hash())?;
+            let next_epoch_start_height =
+                epoch_start_height + epoch_manager.get_epoch_config(epoch_id)?.epoch_length;
+            if next_epoch_start_height <= head.height {
```

Comment:
Under what conditions might this condition *not* be taken? If my reading of the code is correct, we’re iterating through epochs backwards, so the `next_epoch_start_height` should be decreasing and `head.height` should be the highest block height (as the chain progresses forward?)

Reply:
If the requested epoch is the current ongoing epoch, then we can't find its boundary by iterating. 
Added a comments and re-wrote it to check `head.epoch_id` instead of guessing.